api/models/Post.js

```diff
@@ -321,17 +323,80 @@ module.exports = bookshelf.Model.extend(Object.assign({
     return updatedFollowers.concat(newFollowers)
   },
 
+  async addProposalVote ({ userId, optionId }) {
+    return ProposalVote.forge({ post_id: this.id, user_id: userId, option_id: optionId, created_at: new Date() }).save()
+  },
+
   async removeFollowers (usersOrIds, { transacting } = {}) {
     return this.updateFollowers(usersOrIds, { active: false }, { transacting })
   },
 
+  async removeProposalVote ({ userId, optionId }) {
+    const vote = await ProposalVote.query({ where: { user_id: userId, option_id: optionId } }).fetch()
+    return vote.destroy()
+  },
+
+  async setProposalOptions ({ options = [], userId, opts = {} }) {
+    const trxOpts = { require: false, ...opts }
+    return bookshelf.knex.raw(`BEGIN;
+
+      DELETE FROM proposal_options
+      WHERE post_id = ${this.id};
+
+      INSERT INTO proposal_options (post_id, text, color, emoji)
+      VALUES
+        ${options.map(option => `(${this.id}, '${option.text}', '${option.color || ''}', '${option.emoji || ''}')`).join(', ')}
+      RETURNING id;
+
+      COMMIT;`).transacting(trxOpts)
+  },
+
+  async swapProposalVote ({ userId, removeOptionId, addOptionId }) {
+    await this.removeProposalVote({ userId, optionId: removeOptionId })
+    return this.addProposalVote({ userId, optionId: addOptionId })
+  },
+
   async updateFollowers (usersOrIds, attrs, { transacting } = {}) {
-    if (usersOrIds.length == 0) return []
+    if (usersOrIds.length === 0) return []
     const userIds = usersOrIds.map(x => x instanceof User ? x.id : x)
     const existingFollowers = await this.postUsers()
       .query(q => q.whereIn('user_id', userIds)).fetch({ transacting })
     const updatedAttribs = pick(POSTS_USERS_ATTR_UPDATE_WHITELIST, omitBy(isUndefined, attrs))
-    return Promise.map(existingFollowers.models, postUser => postUser.updateAndSave(updatedAttribs, {transacting}))
+    return Promise.map(existingFollowers.models, postUser => postUser.updateAndSave(updatedAttribs, { transacting }))
+  },
+
+  async updateProposalOptions ({ options = [], userId, opts = { } }) {
+    const existingOptions = await this.proposalOptions().fetch({ transaction: opts.transacting, require: false })
+    const existingOptionIds = existingOptions.pluck('id')
+
+    // Delete ALL votes any time options are updated
```

Comment:
couldn't this all happen in setProposalOptions?

Reply:
I could see a case where folks want to add new options after voting has started. people do this for polls in FB all the time. Just an adding a new option wouldn't need to remove all old votes I dont think. but we should probably discuss whether we want this ability or not